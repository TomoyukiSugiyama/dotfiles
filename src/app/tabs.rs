use ratatui::buffer::Buffer;
use ratatui::{layout::Rect, style::Color, style::Style, text::Line, widgets::Widget};
use strum::{Display, EnumIter};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Display, EnumIter)]
pub(crate) enum SelectedTab {
    #[default]
    #[strum(to_string = "Dotfiles")]
    Dotfiles,
    #[strum(to_string = "Execute")]
    Execute,
}

impl SelectedTab {
    pub(crate) fn title(self) -> Line<'static> {
        let line = Line::from(format!("  {self}  "));
        line.style(Style::new().fg(Color::White).bg(Color::Black))
    }

    pub(crate) fn next(self) -> Self {
        match self {
            SelectedTab::Dotfiles => SelectedTab::Execute,
            SelectedTab::Execute => SelectedTab::Dotfiles,
        }
    }
    pub(crate) fn previous(self) -> Self {
        match self {
            SelectedTab::Dotfiles => SelectedTab::Execute,
            SelectedTab::Execute => SelectedTab::Dotfiles,
        }
    }
}

impl Widget for SelectedTab {
    fn render(self, area: Rect, buffer: &mut Buffer) {
        match self {
            SelectedTab::Dotfiles => "Dotfiles".render(area, buffer),
            SelectedTab::Execute => "Execute".render(area, buffer),
        }
    }
}
