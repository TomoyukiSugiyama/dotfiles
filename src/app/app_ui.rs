use super::App;
use super::tabs::SelectedTab;
use ratatui::{
    buffer::Buffer,
    layout::{Constraint, Layout, Rect},
    widgets::{Paragraph, Widget},
};

impl App {
    fn render_title(&mut self, area: Rect, buffer: &mut Buffer) {
        Paragraph::new("Dotfiles Manager")
            .centered()
            .render(area, buffer);
    }

    fn render_inner(&mut self, area: Rect, buffer: &mut Buffer) {
        match self.selected_tab {
            SelectedTab::Dotfiles => self.dotfiles.render(area, buffer),
            SelectedTab::Workflow => self.workflow.render(area, buffer),
        }
    }

    fn render_footer(&mut self, area: Rect, buffer: &mut Buffer) {
        Paragraph::new(
            "Use ←/→ to switch tabs, ↓/↑ to move, Tab to change pane, Enter to run, R to reload config, Home/End to jump, q/Esc to quit.",
        )
            .centered()
            .render(area, buffer);
    }
}

impl Widget for &mut App {
    fn render(self, area: Rect, buffer: &mut Buffer) {
        let [header_area, inner_area, footer_area] = Layout::vertical([
            Constraint::Percentage(5),
            Constraint::Percentage(90),
            Constraint::Percentage(5),
        ])
        .areas(area);

        let [tabs_area, title_area] =
            Layout::horizontal([Constraint::Percentage(70), Constraint::Percentage(30)])
                .areas(header_area);

        self.render_title(title_area, buffer);
        self.selected_tab.render(tabs_area, buffer);
        self.render_inner(inner_area, buffer);
        self.render_footer(footer_area, buffer);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use ratatui::{backend::TestBackend, Terminal};

    fn buffer_to_string(backend: &TestBackend) -> String {
        let buffer = backend.buffer();
        let area = buffer.area();
        let mut result = String::new();

        for y in 0..area.height {
            for x in 0..area.width {
                let cell = buffer.cell((x, y)).expect("valid cell position");
                result.push_str(cell.symbol());
            }
            if y < area.height - 1 {
                result.push('\n');
            }
        }

        result
    }

    #[test]
    fn test_render_app_workflow_tab() {
        let mut app = App::new();
        app.selected_tab = SelectedTab::Workflow;
        
        let mut terminal = Terminal::new(TestBackend::new(100, 30)).unwrap();
        let result = terminal.draw(|frame| frame.render_widget(&mut app, frame.area()));
        
        // Just verify rendering doesn't panic
        assert!(result.is_ok());
    }

    #[test]
    fn test_render_app_dotfiles_tab() {
        let mut app = App::new();
        app.selected_tab = SelectedTab::Dotfiles;
        
        let mut terminal = Terminal::new(TestBackend::new(100, 30)).unwrap();
        let result = terminal.draw(|frame| frame.render_widget(&mut app, frame.area()));
        
        // Just verify rendering doesn't panic
        assert!(result.is_ok());
    }

    #[test]
    fn test_snapshot_app_workflow_tab_with_tools() {
        let mut app = App::new_with_test_tools();
        app.selected_tab = SelectedTab::Workflow;

        let backend = TestBackend::new(120, 40);
        let mut terminal = Terminal::new(backend).unwrap();

        terminal
            .draw(|frame| frame.render_widget(&mut app, frame.area()))
            .unwrap();

        let rendered = buffer_to_string(terminal.backend());
        insta::assert_snapshot!(rendered);
    }

    #[test]
    fn test_snapshot_app_dotfiles_tab_with_tools() {
        let mut app = App::new_with_test_tools();
        app.selected_tab = SelectedTab::Dotfiles;

        let backend = TestBackend::new(120, 40);
        let mut terminal = Terminal::new(backend).unwrap();

        terminal
            .draw(|frame| frame.render_widget(&mut app, frame.area()))
            .unwrap();

        let rendered = buffer_to_string(terminal.backend());
        insta::assert_snapshot!(rendered);
    }
}
