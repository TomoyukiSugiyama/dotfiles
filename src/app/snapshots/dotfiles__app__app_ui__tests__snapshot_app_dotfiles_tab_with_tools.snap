---
source: src/app/app_ui.rs
assertion_line: 128
expression: rendered
---
   Dotfiles   │   Workflow                                                                    Dotfiles Manager          
                                                                                                                        
┌Preferences───────────────────────┐┌Tool Details──────────────────────────────────────────────────────────────────────┐
│Tools Settings                    ││Tool: Brew                                                                        │
│> Brew (brew)                     ││ID: brew                                                                          │
│  Gcloud (gcloud)                 ││Path: /Users/tomoyuki.sugiyama/Work/GitHub/misc/dotfiles/tests/assets/dotfiles/bre│
│  Helm (helm)                     ││Order: Stage 1                                                                    │
│  Krew (krew)                     ││                                                                                  │
│  Rust (rust)                     ││┌Dependency Map (* current tool)─────────────────────────────────────────────────┐│
│  Zsh (zsh)                       │││* Brew (brew)                                                                   ││
│                                  │││|-- - Gcloud (gcloud)                                                           ││
│                                  │││|   `-- - Zsh (zsh)                                                             ││
│                                  │││|-- - Helm (helm)                                                               ││
│                                  │││|   `-- - Zsh (zsh) (repeat)                                                    ││
│                                  │││|-- - Krew (krew)                                                               ││
│                                  │││|   `-- - Zsh (zsh) (repeat)                                                    ││
│                                  │││`-- - Rust (rust)                                                               ││
│                                  │││    `-- - Zsh (zsh) (repeat)                                                    ││
│                                  ││└────────────────────────────────────────────────────────────────────────────────┘│
│                                  ││┌Script──────────────────────────────────────────────────────────────────────────┐│
│                                  │││  #!/usr/bin/env zsh                                                            ││
│                                  │││  echo "brew script"                                                            ││
│                                  │││                                                                                ││
│                                  │││                                                                                ││
│                                  │││                                                                                ││
│                                  │││                                                                                ││
│                                  │││                                                                                ││
│                                  │││                                                                                ││
│                                  │││                                                                                ││
│                                  │││                                                                                ││
│                                  │││                                                                                ││
│                                  │││                                                                                ││
│                                  │││                                                                                ││
│                                  │││                                                                                ││
│                                  │││                                                                                ││
│                                  │││                                                                                ││
│                                  ││└────────────────────────────────────────────────────────────────────────────────┘│
└──────────────────────────────────┘└──────────────────────────────────────────────────────────────────────────────────┘
Use ←/→ to switch tabs, ↓/↑ to move, Tab to change pane, Enter to run, R to reload config, Home/End to jump, q/Esc to qu
